Make autoSkip aware of major ticks

benmccann opened the pull request:
This does the following:
- Skip major ticks only if they can't all fit
- If showing all major ticks, then skip minor ticks such that they're evenly distributed between major ticks
- Remove much of the time scale tick auto generation by generating every tick and then auto-skipping instead. This is so that we don't have two different "auto" algorithms to maintain. Also, by putting it in auto-skip it is available to all scales instead of just one. Thanks @simonbrunel for this suggestion

Interactive examples:
- This PR: https://plnkr.co/edit/REeume?p=preview
- Master: https://plnkr.co/edit/O9FincnNAZVmwtDA1mL9?p=preview

Closes https://github.com/chartjs/Chart.js/issues/4600 https://github.com/chartjs/Chart.js/issues/4612

benmccann commented on src/core/core.scale.js:
```diff
@@ -684,50 +744,65 @@ var Scale = Element.extend({
+		var axisLength = me._axisLength();
+		var tickSize = me._tickSize();
+		var ticksLimit = tickOpts.maxTicksLimit || axisLength / tickSize + 1;
+		var majorIndices = tickOpts.major.enabled ? ticks.flatMap(function(tick, i) {
```
Good catch. Fixed

benmccann commented on src/core/core.scale.js:
```diff
@@ -189,6 +189,66 @@ function parseTickFontOptions(options) {
 	return {minor: minor, major: major};
 }
 
+function factorize(value) {
```
Good thing to check for. Luckily, the input to this function is a number of ticks, so we shouldn't have to worry about values < 1 or values as large as 1E17

benmccann commented:
> Can we have this in a pen/fiddle to test out?

I'm trying out a new one - plnkr :-)  I added a link in the description

kurkle commented:
Seems to work better than master in most cases.
`week` unit does not work well though:
PR:
![image](https://user-images.githubusercontent.com/27971921/57829155-c4bd3c00-77b6-11e9-92ab-06a7a0176e4b.png)

Master:
![image](https://user-images.githubusercontent.com/27971921/57829196-e0284700-77b6-11e9-8f0d-484a62e6a10a.png)

benmccann commented:
Thanks for testing this @kurkle!  I think the issue is actually in the way that the sample was modified to chart `week` data. The PR seems to handle `week` fine in my testing, but the sample wasn't written with that unit in mind. If you just add `week` to the dropdown then it does what you pictured. But that's not quite valid because it will generate major ticks in strange places since the sample uses a custom `afterBuildTicks` which doesn't support weeks. Also, `week` is a unit that isn't auto-selected, so the unit is chosen as `day` and the major unit as `month`, which also breaks the custom `afterBuildTicks` in this sample since it assumes the data is generated daily when the `unit` is day.

Here's a slightly modified sample that works with `week`: https://plnkr.co/edit/jg9aAZ?p=preview

benmccann commented:
I did a little tweaking and got the Code Climate "cognitive complexity" checks passing

kurkle commented:
There is a regression in tick label overlapping:
![image](https://user-images.githubusercontent.com/27971921/58349309-951fcb00-7e6b-11e9-9117-54d025300bdc.png)

![image](https://user-images.githubusercontent.com/27971921/58349550-47579280-7e6c-11e9-9859-7a17fd76e2dd.png)

Quite valid use case IMO, one year, unit=month, maxRotation=0. It does not overlap always.

kurkle commented on test/specs/scale.time.tests.js:
```diff
@@ -127,7 +127,9 @@ describe('Time scale tests', function() {
 			// `bounds === 'data'`: first and last ticks removed since outside the data range
-			expect(ticks).toEqual(['Jan 2', 'Jan 3', 'Jan 4', 'Jan 5', 'Jan 6', 'Jan 7', 'Jan 8', 'Jan 9', 'Jan 10']);
+			expect(ticks.length).toEqual(217);
+			expect(ticks[0]).toEqual('12PM');
```
Why does a time span of 10 days yield `hour` unit? Seems wrong.

benmccann replied:
The scale generates as many ticks as can fit using the smallest unit possible. The time span alone does not determine the unit as the scale width must also be taken into account. If the scale is sufficiently wide then there's room to generate hour labels between each day

benmccann commented on test/specs/global.deprecations.tests.js:
```diff
@@ -333,7 +333,7 @@ describe('Deprecations', function() {
 					type: 'line',
 					data: {
-						labels: ['2015-01-01T20:00:00', '2015-01-01T21:00:00'],
+						labels: ['2015-01-01T20:00:00', '2015-01-01T22:00:00'],
```
changed

benmccann commented on test/specs/core.scale.tests.js:
```diff
@@ -64,14 +60,16 @@ describe('Core.scale', function() {
-			expect(lastTick(chart).label).toBeUndefined();
+			var xAxis = chart.scales['x-axis-0'];
+			var ticks = xAxis.getTicks();
+			expect(ticks[1].label).toBeUndefined();
```
Updated

benmccann commented on src/core/core.scale.js:
```diff
@@ -684,50 +765,50 @@ var Scale = Element.extend({
+		var tickOpts = me.options.ticks;
+		var axisLength = me._axisLength();
+		var ticksLimit = tickOpts.maxTicksLimit || axisLength / me._tickSize() + 1;
+		var majorIndices = tickOpts.major.enabled ? ticks.map(function(tick, i) {
```
done

benmccann commented on src/core/core.scale.js:
```diff
@@ -189,6 +191,85 @@ function parseTickFontOptions(options) {
+		ticksToKeep.push(majorIndices[i]);
+	}
+	for (i = 0; i < ticks.length; i++) {
+		if (ticksToKeep.indexOf(i) < 0) {
```
done

benmccann commented on src/core/core.scale.js:
```diff
@@ -189,6 +189,64 @@ function parseTickFontOptions(options) {
+			result.push(value / i);
+		}
+	}
+	if (Number.isInteger(sqrt)) {
```
done. good catch

benmccann commented on src/core/core.scale.js:
```diff
@@ -684,50 +765,50 @@ var Scale = Element.extend({
 			0;
 	},
 
+	_axisLength: function() {
```
`_autoSkip` was already calculating the axis length, so nothing has changed here, but good to be aware of if we try to make further changes in a follow up PR

benmccann commented:
Thanks for the testing @kurkle. I've fixed the regression. The auto-skipper calculated to keep every 4th day, which was correct. The issue was in the way I changed the `generate` function in the time scale. It was generating a tick on every day, which was incorrect. Since it was a `distribution: 'series'` chart and the sample skipped weekends, it shouldn't have generated ticks on weekends. I've updated it to only generated ticks on days that have data points.

benmccann commented on src/core/core.scale.js:
```diff
@@ -189,6 +191,96 @@ function parseTickFontOptions(options) {
+		ticksToKeep.push(tick);
+	}
+	for (i = Math.max(start, 0); i < end; i++) {
+		if (ticksToKeep.indexOf(i) < 0) {
```
Ok. Have updated this one as well

benmccann commented:
> Too many tests were changed IMO. If "all" ticks are generated and then skipped, then tests should be altered to check for non skipped ones and the result should be quite the same in most cases.

The tests that were changed were mostly tests for `buildTicks`. In these changed tests, only the tick generation is being tested and not the auto-skipping.

benmccann commented:
I'm closing this temporarily. It has a performance regression since it generates more ticks. I will open a new version of this PR soon without the performance regression
